pkg/deploy/strategy/recreate/recreate.go

```diff
@@ -232,3 +241,40 @@ type hookExecutorImpl struct {
 func (i *hookExecutorImpl) Execute(hook *deployapi.LifecycleHook, deployment *kapi.ReplicationController, suffix, label string) error {
 	return i.executeFunc(hook, deployment, suffix, label)
 }
+
+// waitForTerminatedPods waits until all pods for the provided replication controller are terminated.
+func (s *RecreateDeploymentStrategy) waitForTerminatedPods(from *kapi.ReplicationController) error {
+	selector := labels.Set(from.Spec.Selector).AsSelector()
+	options := kapi.ListOptions{LabelSelector: selector}
+	podList, err := s.podClient.Pods(from.Namespace).List(options)
+	if err != nil {
+		return err
+	}
+	// If there are no pods left, we are done.
+	if len(podList.Items) == 0 {
+		return nil
+	}
+	// Watch from the resource version of the list and wait for all pods to be deleted
+	// before proceeding with the Recreate strategy.
+	options.ResourceVersion = podList.ResourceVersion
+	w, err := s.podClient.Pods(from.Namespace).Watch(options)
```

Comment:
1000 watch events across all pods in all namespaces (or if the watch cache is disabled, across all resources in all namespaces)... if dropping this watch fails the deployment, you need to handle re-establishment